Standardise conditional testing fix

tegefaulkes opened the pull request:
### Description
This PR standardises how conditional testing with `testIf` and `describeIf` is used with new reified booleans.

### Issues Fixed

- Fixes #434 
- Related #435 

### Final checklist
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

* [x] Domain specific tests
* [x] Full tests
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

tegefaulkes committed:
test: Moving `utils.ts` into `tests/utils`

Moved `tests/utils.ts` to `tests/utils/utils.ts` and created `test/utils/index.ts`. Trying to keep the utils within one place.

Removed `setupGlobalAgent` code, #420 means it isn't used anymore.

Updating how we handle conditional testing utility's. we're switching to using reified booleans and just composing conditional boolean expression.

#434

tegefaulkes committed:
test: creating reified booleans for conditionally enabling tests

Creating `isPlatformX`, `isTestPlatformX` and `hasX` booleans to use for enabling tests.

Replaced usage of describeIf to use testIf for the tests within the describe block. This means jest shouldn't complain about empty test files if no tests run.

#434

tegefaulkes committed:
test: removing usage of `runTestIfPlatforms` and `runDescribeIfPlatforms`, replaced with `testIf`

We're replacing these with `testIf` and `describeIf` and reified booleans.

#434

CMCDragonkai commented on tests/utils/index.ts:
```diff
@@ -0,0 +1,3 @@
+export * from './utils';
+export * as exec from './exec';
+export * as platform from './platform';
```
I suggest that these are all:

```
export * from './exec';
export * from './platform';
```

This matches how `src/utils/index.ts` does it.

CMCDragonkai replied:
Only need to namespace here if the functions will conflict.

CMCDragonkai commented on tests/nat/endpointIndependentNAT.test.ts:
```diff
@@ -1,19 +1,21 @@
 import Logger, { LogLevel, StreamHandler } from '@matrixai/logger';
 import * as testNatUtils from './utils';
-import { runDescribeIf } from '../utils';
+import { testIf } from '../utils';
```
This should be `import * as testsUtils from '../utils';`. Should be consistent with the rest of usage of utils.

CMCDragonkai commented on tests/nat/endpointIndependentNAT.test.ts:
```diff
@@ -1,19 +1,21 @@
+  hasIptables,
+  hasNsenter,
+  hasUnshare,
+} from '../utils/platform';
```
Yea replace this with just `testsUtils.isPlatformLinux`... etc. Once you have `import * as testsUtils from '../utils';`.

CMCDragonkai commented on tests/utils/platform.ts:
```diff
@@ -0,0 +1,29 @@
+const isTestPlatformMacOs = global.testPlatform === 'macos';
+const isTestPlatformWindows = global.testPlatform === 'windows';
+const isTestPlatformDocker = global.testPlatform === 'docker';
+const isTestPlatformEmpty = global.testPlatform == null;
```
This should be using `globalThis`. And also, can you add a comment above that this is temporary until issue #435 is done. Afterwards these replaced as directories for grouping allowed tests.

CMCDragonkai replied:
Use the docblock comment. Not `//`. Give it whitespace in between like:

```
/**
 * ...
 */

const isTestPlatformDocker ...;
const isTestPlatformMacOS ...;
const isTestPlatformWindows ...;

const isPlatformLinux ...;
```

tegefaulkes committed:
fix: replacing `global` with `globalThis` and adding linting rule

We should be using `globalThis` to access global variables. A linting rule has been added to restrict usage of `global`.

tegefaulkes committed:
test: importing test utils as `testUtils`

CMCDragonkai commented on .eslintrc:
```diff
@@ -28,6 +28,13 @@
+      "error",
+      {
+        "name": "global",
+        "message": "Use globalThis instead"
```
```
"Use `globalThis` instead"
```

tegefaulkes replied:
Updated the message.
